### PR TITLE
Allow using OpenGL ES3 with Angle and extract a Direct3D shared handle

### PR DESCRIPTION
**Pull request by mzanin**
This is a rebased version of #253.

### CODE REVIEW

**Comment by varon** [comment]
Thanks for the PR!

Do you have any test case or something we can use to show this works as expected?

**Comment by mzanin** [comment]
There is a  wpf-gles demo which I have updated to use the rebased dll's so I will push that to a repo. Willlet you know when that is done.

**Comment by l3m** [comment]
@mzanin How about that demo?

**Comment by mzanin** [comment]
Here is the updated demo (equivalent demo as in the previous prq just with the new OpenTK.dll):
[wpf-gles](https://github.com/mzanin/wpf-gles)

File: src/OpenTK/Graphics/GraphicsContext.cs [hunk 1]
```diff
@@ -112,6 +129,18 @@ public GraphicsContext(GraphicsMode mode, IWindowInfo window, int major, int min
                 if (minor < 0)
                     minor = 0;
 
+                // Angle needs an embedded context
+                var use_angle_flag = GraphicsContextFlags.Angle
+                                | GraphicsContextFlags.AngleD3D9
+                                | GraphicsContextFlags.AngleD3D11
+                                | GraphicsContextFlags.AngleOpenGL;
+                var use_angle = false;
+                if ((flags & use_angle_flag) != 0)
```
Comment (varon) [inline src/OpenTK/Graphics/GraphicsContext.cs]: The implicit addition of the embedded context flag is a nice touch here.

**Comment by varon** [comment]
I also definitely have to add: thanks for the contribution; it's always hugely appreciated to get nice additions like this to the library.

File: src/OpenTK/Graphics/GraphicsContext.cs (hunk 1 above)
Comment (mzanin) [inline src/OpenTK/Graphics/GraphicsContext.cs]: I don't have any issues with changing the naming convention for the code I introduced but I just wanted to make you aware that there are a lot of pre-existing variable names which are named according to the `snake_style ` convention, even just in the constructor in GraphicsContext.cs which you commented on above (i.e. `handle_cached`). In fact there seems to be a discrepancy as I see some locally defined variable names using `camelCase` whereas a lot of member variable names are `snake_style`. Is this a rule?

Comment (varon) [inline src/OpenTK/Graphics/GraphicsContext.cs]: That's a totally reasonable point. The best explanation of the situation I can give is that OpenTK has been around for ages. Back when it first started, which is where a lot of this older code is from, it's very likely that the conventions for the library were different or just less rigorous.

I'm not sure at which point OpenTK adopted the guidelines, but it's pretty clearly stated in the [wiki's contributing page](https://github.com/opentk/opentk/wiki/Contributing).
Specifically, the current [Framework Design Guidelines](https://msdn.microsoft.com/en-us/library/ms229045(v=vs.110).aspx) state:
>X DO NOT use underscores, hyphens, or any other nonalphanumeric characters.

Thanks for your patience on this. I really do sympathise with the general annoyance of trying to meet an unclear convention in a PR.

Comment (mzanin) [inline src/OpenTK/Graphics/GraphicsContext.cs]: Ok no worries, I pushed the changes you requested.

Comment (varon) [inline src/OpenTK/Graphics/GraphicsContext.cs]: Thanks!

**Comment by varon** [comment]
Thanks again for your hard work on this, @mzanin!

**Comment by l3m** [comment]
Haha, nice! Who would have thought that this would finally get accepted :)

**Comment by JimSEOW** [comment]
I would like to do OpenGL ES Win10 UWP XAML interop for c#
For Angle.WindowStore, there is a c++ template and a WIP c# template.
For OpenTK, what is the latest status for Win10 UWP XAML ? I only notice WPF XAML interop.
There is a recent [discussion](https://github.com/Microsoft/HoloJS/issues/57#issuecomment-314534809).

**Pull request merged**